lib/ecto/adapters/mssql/connection.ex

```diff
@@ -0,0 +1,1644 @@
+      do: {acc, user_acc}
+
+    defp intersperse_reduce([elem], _separator, user_acc, reducer, acc) do
+      {elem, user_acc} = reducer.(elem, user_acc)
+      {[acc | elem], user_acc}
+    end
+
+    defp intersperse_reduce([elem | rest], separator, user_acc, reducer, acc) do
+      {elem, user_acc} = reducer.(elem, user_acc)
+      intersperse_reduce(rest, separator, user_acc, reducer, [acc, elem, separator])
+    end
+
+    defp if_do(condition, value) do
+      if condition, do: value, else: []
+    end
+
+    defp escape_string(value) when is_binary(value) do
+      value
+      |> :binary.replace("'", "''", [:global])
+    end
+
+    defp ecto_cast_to_db(type, opts) do
+      size = Keyword.get(opts, :size)
+      precision = Keyword.get(opts, :precision)
+      scale = Keyword.get(opts, :scale)
+      ecto_to_db(type, size, precision, scale)
+    end
+
+    defp ecto_to_db(type, size, precision, scale, query \\ nil)
+
+    defp ecto_to_db({:array, _}, _, _, _, query),
+      do: error!(query, "Array type is not supported by TDS")
+
+    defp ecto_to_db(:id, _, _, _, _), do: "bigint"
+    defp ecto_to_db(:integer, _, _, _, _), do: "integer"
+    defp ecto_to_db(:bigint, _, _, _, _), do: "bigint"
+    defp ecto_to_db(:serial, _, _, _, _), do: "int IDENTITY(1,1)"
+    defp ecto_to_db(:bigserial, _, _, _, _), do: "bigint IDENTITY(1,1)"
+    defp ecto_to_db(:binary_id, _, _, _, _), do: "uniqueidentifier"
+    defp ecto_to_db(:boolean, _, _, _, _), do: "bit"
+    defp ecto_to_db(:string, nil, _, _, _), do: "nvarchar(255)"
+    defp ecto_to_db(:string, :max, _, _, _), do: "nvarchar(max)"
+    defp ecto_to_db(:string, s, _, _, _) when s in 1..4_000, do: "nvarchar(#{s})"
+    defp ecto_to_db(:float, nil, _, _, _), do: "float"
+    defp ecto_to_db(:float, s, _, _, _) when s in 1..53, do: "float(#{s})"
+    defp ecto_to_db(:binary, nil, _, _, _), do: "varbinary(2000)"
+    defp ecto_to_db(:binary, :max, _, _, _), do: "varbinary(max)"
+    defp ecto_to_db(:binary, s, _, _, _) when s in 1..8_000, do: "varbinary(#{s})"
+    defp ecto_to_db(:uuid, _, _, _, _), do: "uniqueidentifier"
+    defp ecto_to_db(:map, nil, _, _, _), do: "nvarchar(2000)"
+    defp ecto_to_db(:map, :max, _, _, _), do: "nvarchar(max)"
+    defp ecto_to_db(:map, s, _, _, _) when s in 0..4_000, do: "nvarchar(#{s})"
+    defp ecto_to_db({:map, _}, nil, _, _, _), do: "nvarchar(2000)"
+    defp ecto_to_db({:map, _}, s, _, _, _) when s in 1..4_000, do: "nvarchar(#{s})"
+    defp ecto_to_db({:map, _}, :max, _, _, _), do: "nvarchar(max)"
+    defp ecto_to_db(:time, _, _, _, _), do: "time(0)"
+    defp ecto_to_db(:time_usec, _, p, _, _) when p in 0..7, do: "time(#{p}})"
+    defp ecto_to_db(:time_usec, _, _, _, _), do: "time(6)"
+    defp ecto_to_db(:utc_datetime, _, _, _, _), do: "datetimeoffset(0)"
+    defp ecto_to_db(:utc_datetime_usec, _, p, _, _) when p in 0..7, do: "datetimeoffset(#{p})"
+    defp ecto_to_db(:utc_datetime_usec, _, _, _, _), do: "datetimeoffset(6)"
```

Comment:
`:utc_datetime` is still a datetime without offset in other adapters, as we don't want to handle the offset in the database, rather on Ecto side. So we should use the same as the naive datetime types below:

```suggestion
    defp ecto_to_db(:utc_datetime, _, _, _, _), do: "datetime"
    defp ecto_to_db(:utc_datetime_usec, _, p, _, _) when p in 0..7, do: "datetime2(#{p})"
    defp ecto_to_db(:utc_datetime_usec, _, _, _, _), do: "datetime2(6)"
```

Reply:
is there anything in ecto that will convert NaiveDateTime to DateTime out of the box when UTC field is used? There is no way that driver would know this if there is no custom loader or dumper that I removed btw.

It also looks kind of redundant to use both DateTime and NaiveDateTime since they are mapped to same db type

Reply:
The adapter can map both DateTime and NaiveDateTime to datetime2 when sending parameters. When reading the parameters, you can always return NaiveDateTime and Ecto will cast it accordingly. If for some reason this change does not work, please let me know.